modules/imgproc/src/contours_common.hpp

```diff
@@ -135,47 +297,140 @@ template <typename T>
 class TreeIterator
 {
 public:
-    TreeIterator(Tree<T>& tree_) : tree(tree_)
+    TreeIterator(Tree<T>& tree_) : tree(tree_)//,levels(&tree._treeIteratorArena)
     {
         CV_Assert(!tree.isEmpty());
-        levels.push(0);
+        levels.push_back(0);
     }
     bool isDone() const
     {
         return levels.empty();
     }
     const TreeNode<T>& getNext_s()
     {
-        int idx = levels.top();
-        levels.pop();
+        int idx = levels.back();
+        levels.pop_back();
         const TreeNode<T>& res = tree.elem(idx);
         int cur = tree.lastSibling(res.first_child);
         while (cur != -1)
         {
-            levels.push(cur);
+            levels.push_back(cur);
             cur = tree.elem(cur).prev;
         }
         return res;
     }
 
 private:
-    std::stack<int> levels;
     Tree<T>& tree;
+    vectorOfRanges<int> levels;
 };
 
 //==============================================================================
 
+class ContourPointsStorage
```

Comment:
Can we combine `ContourPointsStorage` and `ContourCodesStorage` into a single template?